sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/windowing/WindowMappingFn.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.beam.sdk.transforms.windowing;
+
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.values.PCollectionView;
+import org.joda.time.Duration;
+
+/**
+ * A function that takes the windows of elements in a main input and maps them to the appropriate
+ * window in a {@link PCollectionView} consumed as a
+ * {@link ParDo.BoundMulti#withSideInputs(PCollectionView[]) side input}.
+ */
+public abstract class WindowMappingFn<W extends BoundedWindow> {
```

Comment:
It would actually be perfectly fine to have a `WindowMappingFn` that could only apply to certain main input window types and side input window types. But due to erasure we can't actually validate the main window type or side window type, unless we play tricks like we do with other Fns. I don't think it is worthwhile here so I don't see the point of either one (conversely, I'm perfectly happy having both and leaving them as wildcards / `BoundedWindow` when a runner interacts with them)

The feature as it exists before this PR probably has ~zero users and may have just barely more afterwards. This tweak in the design is more valuable for clarifying the model and solving the GC issue that real use cases.

I was thinking of two things:

1. The type variables for internal type checking of implementations, which you can just do with their own type variables that don't show up in the interface.
2. Just accurately describing the situation since it has no downside that I can think of and tends to pay off later in unforeseen ways.